Mathlib/LinearAlgebra/FreeModule/PID.lean

```diff
@@ -440,6 +441,65 @@ structure Basis.SmithNormalForm (N : Submodule R M) (ι : Type*) (n : ℕ) where
   snf : ∀ i, (bN i : M) = a i • bM (f i)
 #align basis.smith_normal_form Basis.SmithNormalForm
 
+namespace Basis.SmithNormalForm
+
+variable {n : ℕ} {N : Submodule R M} (snf : Basis.SmithNormalForm N ι n) (m : N)
+
+lemma repr_eq_zero_of_nmem_range {i : ι} (hi : i ∉ Set.range snf.f) :
+    snf.bM.repr m i = 0 := by
+  obtain ⟨m, hm⟩ := m
+  obtain ⟨c, rfl⟩ := snf.bN.mem_submodule_iff.mp hm
+  replace hi : ∀ j, snf.f j ≠ i := by simpa using hi
+  simp [Finsupp.single_apply, hi, snf.snf]
+
+@[simp] lemma repr_comp_embedding_eq_smul :
+    snf.bM.repr m ∘ snf.f = (snf.bN.repr m : Fin n → R) • snf.a := by
```

Comment:
I'm surprised that this takes so much work, since we already have the field `snf : ∀ (i : Fin n), ↑(↑bN i) = a i • ↑bM (↑f i)`. Maybe there's some general theorem about bases waiting to be extracted? (I can try doing so tomorrow.)

Reply:
I don't see a result to extract but of course it would be great if you find one!

Also this is only a 12-line proof and there is some genuine content here.

Reply:
I still think there is something to be extracted here, but about an hour of trying didn't give me anything sensible. So it's probably not worth trying further right now.